compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs

```diff
@@ -86,6 +86,35 @@ fn expand_or_pat<'p, 'tcx>(pat: &'p Pat<'tcx>) -> Vec<&'p Pat<'tcx>> {
     pats
 }
 
+/// Evaluate an int constant, with a faster branch for a common case.
+#[inline]
+fn fast_try_eval_bits<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    value: &mir::Const<'tcx>,
+) -> Option<u128> {
+    let int  = match value {
+        // If the constant is already evaluated, we shortcut here.
+        mir::Const::Ty(c) if let ty::ConstKind::Value(valtree) = c.kind() => {
+            valtree.unwrap_leaf()
+        },
+        // This is a more general form of the previous case.
+        _ => {
+            value.try_eval_scalar_int(tcx, param_env)?
+        },
+    };
```

Comment:
Can this optimization be done in the original `try_eval_scalar`?

Reply:
Actually let me first check it's a useful optimization, it's been there for a while

Reply:
Ok, I moved it to `try_eval_scalar_int`. I could move it to `try_eval_scalar`, but I would need to be sure it's only called on types that have leaf-only valtrees. Can I be sure of that?

Reply:
You cannot. It could be called on references, that have a branch valtree, that is turned into a pointer scalar. 
I had not seen that comment before approving. Please r- if you want to change it.

Reply:
`try_eval_scalar_int` is the right place to put it then, all good

Reply:
References are not a branch valtree, they are using the same valtree as what they point to.

However that's still the wrong result since an `&i32` would return the scalar integer.